### PR TITLE
CheckIfTypeIsEquivalent improvement

### PR DESCRIPTION
**Pull request by kunalspathak**
* Always emit a bailout instead of helper if Equivalent Type check fails
* If equivalent type check is full, start evicting non-shared types.

### CODE REVIEW

**Comment by kunalspathak** [comment]
@pleath , @LouisLaf  , please review!

File: lib/Backend/BackwardPass.cpp [hunk 1]
```diff
@@ -2107,6 +2107,16 @@ BackwardPass::DeadStoreTypeCheckBailOut(IR::Instr * instr)
         return;
     }
 
+    // If oldBailoutKind is equivTypeCheck and appropriate opnd doesn't have initial/final type
+    // then leave alone the bailout
+    if (oldBailOutKind == IR::BailOutFailedEquivalentTypeCheck || oldBailOutKind == IR::BailOutFailedEquivalentFixedFieldTypeCheck)
+    {
+        if (!(propertySymOpnd->HasInitialType() || propertySymOpnd->HasFinalType()))
```
Comment (pleath) [inline lib/Backend/BackwardPass.cpp]: Is this part necessary? If we're really emitting an equivalent bailout, then final type optimization isn't involved, right?

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 2]
```diff
@@ -8243,6 +8243,20 @@ namespace Js
             return true;
         }
 
+        // Since the equivTypes cache is always compacted, if last entry is non-nullptr, it means
+        // the cache is full.
+        DynamicType * dynamicType = DynamicType::Is(type->GetTypeId()) ? static_cast<DynamicType*>(type) : nullptr;
+        bool isEquivTypesCacheIsFull = equivTypes[EQUIVALENT_TYPE_CACHE_SIZE - 1] != nullptr;
```
Comment (pleath) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Nit: remove one "is" from the name?

File: lib/Backend/BackwardPass.cpp (hunk 1 above)
Comment (kunalspathak) [inline lib/Backend/BackwardPass.cpp]: That's what I thought, but we do end up here with non-null initialType and finalType and if I don't have this, we hit assert in Lower where that if it is not equivTypeCheck then there shouldn't be equiv bailouts.
```c++
Assert(doEquivTypeCheck || !instrChk->HasEquivalentTypeCheckBailOut());
```

File: lib/Runtime/Language/JavascriptOperators.cpp (hunk 2 above)
Comment (kunalspathak) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Ah yes. Will do.

File: lib/Backend/BackwardPass.cpp (hunk 1 above)
Comment (pleath) [inline lib/Backend/BackwardPass.cpp]: If that's the case, then can you find out how we got here with an equiv bailout kind?

Comment (kunalspathak) [inline lib/Backend/BackwardPass.cpp]: Sure.

Comment (kunalspathak) [inline lib/Backend/BackwardPass.cpp]: So I was mistakenly checking the `bailoutKind` that we have in forward pass instead of checking that for backward pass. With that, I don't need the check for initial/final type.

File: lib/Backend/Lower.cpp [hunk 3]
```diff
@@ -5710,9 +5710,6 @@ Lowerer::GenerateLdFldWithCachedType(IR::Instr * instrLdFld, bool* continueAsHel
 
     if (hasTypeCheckBailout)
     {
-        AssertMsg(PHASE_ON1(Js::ObjTypeSpecIsolatedFldOpsWithBailOutPhase) || !propertySymOpnd->IsTypeDead(),
```
Comment (pleath) [inline lib/Backend/Lower.cpp]: Can you leave the asserts but assert that either the type is dead or the bailout kind is equivalent?

Comment (kunalspathak) [inline lib/Backend/Lower.cpp]: Turned out that it is fine to leave assert here and the one below because we will only come this code path if there is no typeCheckBailout.
```c++
if(hasTypeCheckBailout) {

```

File: lib/Backend/BackwardPass.cpp [hunk 4]
```diff
@@ -2073,8 +2073,8 @@ BackwardPass::DeadStoreTypeCheckBailOut(IR::Instr * instr)
             preOpBailOutInstrToProcess = nullptr;
         }
         return;
-    }
-
+    } 
+
```
Comment (dilijev) [inline lib/Backend/BackwardPass.cpp]: nit: trailing whitespace

File: lib/Backend/Lower.cpp [hunk 5]
```diff
@@ -6825,8 +6824,7 @@ Lowerer::GenerateStFldWithCachedType(IR::Instr *instrStFld, bool* continueAsHelp
...
-            "Why does a field store have a type check bailout, if its type is dead?");
+        AssertMsg(PHASE_ON1(Js::ObjTypeSpecIsolatedFldOpsWithBailOutPhase) || !propertySymOpnd->IsTypeDead(), "Why does a field load have a type check bailout, if its type is dead?");
```
Comment (dilijev) [inline lib/Backend/Lower.cpp]: In this line and above, it looks like there's no change. Can you restore to the original version? (splitting into 2 lines will be slightly easier to read)

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 6]
```diff
@@ -8243,6 +8243,20 @@ namespace Js
             return true;
         }
 
+        // Since the equivTypes cache is always compacted, if last entry is non-nullptr, it means
+        // the cache is full.
```
Comment (dilijev) [inline lib/Runtime/Language/JavascriptOperators.cpp]: nit: no need to break this comment to multiple lines since the code following is longer.

File: lib/Backend/Lower.cpp (hunk 5 above)
Comment (dilijev) [inline lib/Backend/Lower.cpp]: Correction -- the one above says "load" and this one originally said "store" but now says "load" -- was that change intentional?

---
In reply to: [79080126](https://github.com/Microsoft/ChakraCore/pull/1578#discussion_r79080126) [](ancestors = 79080126)
